daffodil-cli/src/main/scala/org/apache/daffodil/cli/Main.scala

```diff
@@ -1165,13 +1168,24 @@ class Main(
           case Some(processor) => {
             Assert.invariant(!processor.isError)
             val input = parseOpts.infile.toOption match {
-              case Some("-") | None => STDIN
+              case Some("-") | None => InputSourceDataInputStream(STDIN)
               case Some(file) => {
-                val f = new File(file)
-                new FileInputStream(f)
+                // for files <= 2GB, use a mapped byte buffer to avoid the overhead related to
+                // the BucketingInputSource. Larger files cannot be mapped so we cannot avoid it
+                val path = Paths.get(file)
+                val size = Files.size(path)
+                if (size <= Int.MaxValue) {
```

Comment:
doc for the map method says: 

    For most operating systems, mapping a file into memory is more expensive than reading or writing a 
    few tens of kilobytes of data via the usual read

So should there be a floor check also e.g., below some size we just read it into a byte buffer and avoid the map?

Reply:
Possibly. Though, I imagine if you're parsing a small file with the CLI then the overhead of mmap is going to be relatively small compared to the overhead of starting up a JVM and maybe the it won't make a difference? I'm not sure. We can do some experiments to see if there's a benefit for smaller files.

Reply:
Could leave this as you have it, and we can look at nightly performance stuff to see if it slows down noticably. Lots of those will use files that are small.

Reply:
The nightlies don't use the `parse` command so won't see any change. They use the `performance` command which reads test files into a byte array before testing to avoid overhead related to disk I/O.

We could create some patches that run on the nightlies, one patch change the performance command to use FileInputStream and one to use a MappedByteBuffer, which would give us an idea of mmap vs file input stream. But that's feels like a decent amount of work just to figure out an optimal size where mmap overhead > bucketing overhead. Also, based on my bucketing vs non-bucketing tests, I feel like bucketing overhead is probably more than mmap-overhead, even with small files and so we should always avoid bucketing when possible.